### ISSUE TITLE
how to get same session_id from multiple server?

### ISSUE THREAD

**Issue by LeeMinKyu** [spring-projects/spring-session#143]
I'm sorry that my english is still not so good.
Please bear with me, I hope you can understand my question..

---

I have two web servers.
(each web application is same)

Web servers are sharing one redis server.

And I use Spring Security and Spring Session.

When I login first server and access second server,
I want to login second server automatically, but it isn't.

I guess,
because session id is different from different server ip.

* how to get same session id ?

It is server1.
![image](https://cloud.githubusercontent.com/assets/8723117/6140714/055d4b1c-b1df-11e4-84c7-5d2b3aaf0989.png)

It is server2.
![image](https://cloud.githubusercontent.com/assets/8723117/6140734/4eed2b44-b1df-11e4-8bfe-27f0cb4494b9.png)

 
WEB.XML
----

	
	
	<context-param>
		<param-name>contextConfigLocation</param-name>
		<param-value>
			/WEB-INF/spring/root-context.xml,
			/WEB-INF/spring/spring-security.xml,
			/WEB-INF/spring/jedis.xml
                </param-value>
	</context-param>
	
	
	<listener>
		<listener-class>org.springframework.web.context.ContextLoaderListener</listener-class>
	</listener>

	
	<servlet>
		<servlet-name>appServlet</servlet-name>
		<servlet-class>org.springframework.web.servlet.DispatcherServlet</servlet-class>
		<init-param>
			<param-name>contextConfigLocation</param-name>
			<param-value>/WEB-INF/spring/appServlet/servlet-context.xml</param-value>
		</init-param>
		<load-on-startup>1</load-on-startup>
	</servlet>

	<servlet-mapping>
		<servlet-name>appServlet</servlet-name>
		<url-pattern>/</url-pattern>
	</servlet-mapping>

	
	<filter>
		<filter-name>CharacterEncodingFilter</filter-name>
		<filter-class>org.springframework.web.filter.CharacterEncodingFilter</filter-class>
		<init-param>
			<param-name>encoding</param-name>
			<param-value>utf-8</param-value>
		</init-param>
		<init-param>
			<param-name>forceEncoding</param-name>
			<param-value>true</param-value>
		</init-param>
	</filter>
	<filter-mapping>
		<filter-name>CharacterEncodingFilter</filter-name>
		<url-pattern>/*</url-pattern>
	</filter-mapping>

	
 	<filter>
		<filter-name>springSessionRepositoryFilter</filter-name>
		<filter-class>org.springframework.web.filter.DelegatingFilterProxy</filter-class>
	</filter>
	<filter-mapping>
		<filter-name>springSessionRepositoryFilter</filter-name>
		<url-pattern>/*</url-pattern>
	</filter-mapping>

	
	<filter>
		<filter-name>springSecurityFilterChain</filter-name>
		<filter-class>org.springframework.web.filter.DelegatingFilterProxy</filter-class>
	</filter>
	<filter-mapping>
		<filter-name>springSecurityFilterChain</filter-name>
		<url-pattern>/*</url-pattern>
	</filter-mapping>

jedis.xml
---

	<bean id="jedisConnectionFactory" class="org.springframework.data.redis.connection.jedis.JedisConnectionFactory">
		<property name="hostName" value="" />
		<property name="port" value="6379" />
		<property name="poolConfig" ref="redisPoolConfig" />
	</bean>
	
	
	<bean id="redisPoolConfig" class="redis.clients.jedis.JedisPoolConfig">
		<property name="testOnBorrow" value="true" />
		<property name="minEvictableIdleTimeMillis" value="60000" />
		<property name="softMinEvictableIdleTimeMillis" value="1800000" />
		<property name="numTestsPerEvictionRun" value="-1" />
		<property name="testOnReturn" value="false" />
		<property name="testWhileIdle" value="true" />
		<property name="timeBetweenEvictionRunsMillis" value="30000" />
	</bean>
	
	
	<bean id="stringRedisSerializer"
		class="org.springframework.data.redis.serializer.StringRedisSerializer" />
		
	
	<bean id="redisTemplate" class="org.springframework.data.redis.core.RedisTemplate" >
		<property name="connectionFactory" ref="jedisConnectionFactory" />
		<property name="keySerializer" ref="stringRedisSerializer" />
		<property name="hashKeySerializer" ref="stringRedisSerializer" />
	</bean>

spring-security.xml
---
	<http pattern="/resources/**" security="none" />

	<http auto-config="true" >
		<session-management session-fixation-protection="changeSessionId">
			<concurrency-control max-sessions="1" error-if-maximum-exceeded="true"/>
		</session-management>
		<intercept-url pattern="/" access="ROLE_ANONYMOUS, ROLE_USER" />
		<intercept-url pattern="/perBoard" access="ROLE_ANONYMOUS, ROLE_USER" />
		<intercept-url pattern="/per" access="ROLE_ANONYMOUS, ROLE_USER" />
		<intercept-url pattern="/perSearchTag" access="ROLE_ANONYMOUS, ROLE_USER" />
		<intercept-url pattern="/**" access="ROLE_USER" />
		<form-login login-page="/" 
					authentication-success-handler-ref="loginSuccessHandler"
					authentication-failure-handler-ref="loginFailureHandler"
					always-use-default-target="true" 
					username-parameter="j_username" 
					password-parameter="j_password"/>
					
		<logout logout-success-url="/" invalidate-session="true" delete-cookies="true" />
	</http>

	<authentication-manager>
		<authentication-provider user-service-ref="userDetailsService" />
	</authentication-manager>
    
 	<beans:bean id="loginSuccessHandler" class=".......LoginSuccessHandler">
 		<beans:property name="sqlSession" ref="sqlSession" />
 	</beans:bean>
 	<beans:bean id="loginFailureHandler" class=".......LoginFailureHandler" />
	<beans:bean id="userDetailsService" class="......UserDetailsServiceImpl">
		<beans:property name="sqlSession" ref="sqlSession" />
	</beans:bean>

**Comment by Thilak-T** [comment]
Just want to know whether do you login to first server from one browser and access second server from another browser?

In your config jedis.xml, there is no IP for redis server. does it defaults to localhost?

**Comment by rwinch** [comment]
@LeeMinKyu Thank you for the detailed explanation. Are the applications hosted on the same host port and context path? If not, the cookie will be isolated for each application instance.

If that doesn't help...Can you post the HTTP Response for when the cookies is set for each application?

**Comment by LeeMinKyu** [comment]
@Thilak-T Thank you for your comment.
Browser is Chrome and no access from another browser.
Redis server ip is a specific ip. I just think that redis server ip doesn't have to expose on this issue. 

@rwinch  Thank you for your comment.
Web applications are same(port, path, os version, apache, tomcat version, jdk version ... all same ).

Server 1 request and response  - login page ( == home )
![image](https://cloud.githubusercontent.com/assets/8723117/6182933/cdeef3fc-b390-11e4-873b-148c0fdca534.png)

Server 2 request and response  - login page ( == home )
![image](https://cloud.githubusercontent.com/assets/8723117/6182941/021c4e36-b391-11e4-8c3f-85000b13afcc.png)

In this case, Is it impossible to get same session id ??
If host of request header becomes equal, can I get the same session id ?

---
<br>
**Additionally**
I find that if I access same server from different URLs , session id in cookie is not equal.

I access my local server from http://localhost/
![image](https://cloud.githubusercontent.com/assets/8723117/6183036/d4623ce2-b392-11e4-91d9-5c09a61be3c6.png)

access my local server from http://127.0.0.1/
![image](https://cloud.githubusercontent.com/assets/8723117/6183046/054d9aa4-b393-11e4-8c18-0c550f98941b.png)

access my local server from private ip
![image](https://cloud.githubusercontent.com/assets/8723117/6183061/3ae50a94-b393-11e4-9092-b69bfd583063.png)

Is it intended?
Session id is affected by domain url? Can I customize session id?

**Comment by rwinch** [comment]
The request/responses were not of the SESSION cookie getting set (the SESSION cookie was already present. Please submit the request/responses when the cookie was first set (hint clear out the cookies).

> I find that if I access same server from different URLs , session id in cookie is not equal.

Yes this is how browsers work. If you get a session cookie from yourbank.com you do not want evil.com to be able to read or write to yourbank.com's cookies.

**Comment by LeeMinKyu** [comment]
@rwinch  Thank you for your comment.
And I'm so sorry. At last, I can understand that ' ***when the cookie was first set*** '.
You want that the response headers have 'Set-Cookie' attribute,, right?
I followed your comment, cleared out the cookies and re-accessed my web pages.

**Server1**

---
![image](https://cloud.githubusercontent.com/assets/8723117/6203067/2c200b2c-b54f-11e4-836b-2e1f8fbdfa41.png)


**Server2**

---
![image](https://cloud.githubusercontent.com/assets/8723117/6203063/1ad17b08-b54f-11e4-8f80-c9e8fa9c73ff.png)


I guess, 
If the request URLs are same, the session cookie's values are same even though each network headers's remote address( = server IP ) is different.
Is it right?
If right, my web.xml, spring-security.xml and jedis.xml  are correct? 
If not, How do I set the spring - mvc + security + session + redis for sharing session in multiple servers?
I want to share a session for sharing user's login_info and auth in multiple servers.

Once again, thank you for your comments and sorry about my not good english.
